ePBFT/src/pbft/libbyz/Replica.cpp

```diff
@@ -229,13 +229,101 @@ void Replica::receive_message(const uint8_t* data, uint32_t size)
   }
 }
 
+bool Replica::compare_execution_results(
+  const ByzInfo& info, Pre_prepare* pre_prepare)
+{
+  auto& pp_root = pre_prepare->get_merkle_root();
+  if (!std::equal(
+        std::begin(pp_root), std::end(pp_root), std::begin(info.merkle_root)))
+  {
+    LOG_FAIL << "Merkle root between execution and the pre_prepare message "
+                "does not match, seqno:"
+             << pre_prepare->seqno() << std::endl;
+    return false;
+  }
+
+  auto tx_ctx = pre_prepare->get_ctx();
+  if (tx_ctx != info.ctx)
+  {
+    LOG_FAIL << "User ctx between execution and the pre_prepare message "
+                "does not match, seqno:"
+             << pre_prepare->seqno() << std::endl;
+    return false;
+  }
+  return true;
+}
+
+size_t Replica::ledger_cursor() const
+{
+  return ledger_replay->cursor();
+}
+
+bool Replica::apply_ledger_data(const std::vector<uint8_t>& data)
+{
+  PBFT_ASSERT(
+    !data.empty(), "apply ledger data should not receive empty vector");
+
+  auto executable_pp =
+    ledger_replay->apply_data(data, rqueue, brt, ledger_writer.get());
+
+  if (executable_pp)
+  {
+    auto pre_prepare = executable_pp.get();
```

Comment:
Please do not create this variable if you have the std::unique_ptr and implies ownership properties, you should not need to undermine them in the same function